### ISSUE TITLE
git repo fails to clone then r10k purges everything breaking the masters

### ISSUE THREAD

**Issue by kholloway** [puppetlabs/r10k#178]
Occasionally my git repo fails to clone for some reason, when that happens r10k then purges the environments directories and wipes out everything, our puppet masters then fail to properly work.

Seems like the destination directory should be left alone if r10k fails to run to completion for all repos BEFORE wiping out what was a working directory/setup.

Is there any current CLI option I'm missing to fix this?

Thanks!
### R10k commands tried:

`r10k deploy environment -pv`
`r10k deploy environment`
### R10k CLI output

```
[R10K::Source::Git - INFO] Determining current branches for "git@gitlab.name.changed:teamname/puppet3.git"
[R10K::TaskRunner - ERROR] Task #<R10K::Task::Deployment::DeployEnvironments:0x00000001dffdf8> failed while running: Couldn't update git cache for git@
gitlab.name.changed:teamname/puppet3.git: "fatal: Could not read from remote repository."
[R10K::Task::Deployment::PurgeEnvironments - INFO] Purging stale environments from /etc/puppet/environments
```
### /etc/r10k.yaml

```
# The location to use for storing cached Git repos
:cachedir: '/var/cache/r10k'

# A list of git repositories to create
:sources:
  # This will clone the git repository and instantiate an environment per
  # branch in /etc/puppet/environments
  :puppet3:
    remote: 'git@gitlab.name.changed:teamname/puppet3.git'
    basedir: '/etc/puppet/environments'
```
### Puppetfile from repo puppet3.git

```
forge 'forge.puppetlabs.com'

## Specify Forge modules here
mod 'puppetlabs/vcsrepo', '1.1.0'
mod 'puppetlabs/stdlib', '4.3.2'
mod 'puppetlabs/concat', '1.0.4'

# Specify any local Gitlab repos here
# Old Puppet repo modules - by default master will be checked out
mod "old_puppet_repo",
   :git => 'git@gitlab.name.changed:teamname/puppet.git'
```

**Comment by adrienthebo** [comment]
Well, that's something I didn't expect to see. Could you run `r10k deploy environment -v debug2 --trace` and try to reproduce this with backtraces and the higher logging?

**Comment by kholloway** [comment]
Below is the output from the debug, our gitlab server fails the connection for some reason followed by R10k purging everything. As soon as the gitlab server works fine again then everything comes back again as expected.

```
[R10K::Source::Git - INFO] Determining current branches for "git@gitlab.name.changed:teamname/puppet3.git"
[R10K::Git::Cache - DEBUG1] Execute: git --git-dir /var/cache/r10k/git@gitlab.name.changed-teamname-puppet3.git fetch --prune origin
[R10K::Git::Cache - DEBUG2] [git --git-dir /var/cache/r10k/git@gitlab.name.changed-teamname-puppet3.git fetch --prune origin] STDERR: ssh_exchange_identification: Connection closed by remote host
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
[R10K::TaskRunner - ERROR] Task #<R10K::Task::Deployment::DeployEnvironments:0x00000001099718> failed while running: Couldn't update git cache for git@gitlab.name.changed:teamname/puppet3.git: "fatal: Could not read from remote repository."
/usr/lib64/ruby/gems/2.0.0/gems/r10k-1.3.2/lib/r10k/git/cache.rb:73:in `rescue in sync!'
/usr/lib64/ruby/gems/2.0.0/gems/r10k-1.3.2/lib/r10k/git/cache.rb:58:in `sync!'
/usr/lib64/ruby/gems/2.0.0/gems/r10k-1.3.2/lib/r10k/git/cache.rb:52:in `sync'
/usr/lib64/ruby/gems/2.0.0/gems/r10k-1.3.2/lib/r10k/source/git.rb:67:in `preload!'
/usr/lib64/ruby/gems/2.0.0/gems/r10k-1.3.2/lib/r10k/deployment.rb:28:in `each'
/usr/lib64/ruby/gems/2.0.0/gems/r10k-1.3.2/lib/r10k/deployment.rb:28:in `preload!'
/usr/lib64/ruby/gems/2.0.0/gems/r10k-1.3.2/lib/r10k/task/deployment.rb:76:in `call'
/usr/lib64/ruby/gems/2.0.0/gems/r10k-1.3.2/lib/r10k/task_runner.rb:26:in `block in run'
/usr/lib64/ruby/gems/2.0.0/gems/r10k-1.3.2/lib/r10k/task_runner.rb:21:in `catch'
/usr/lib64/ruby/gems/2.0.0/gems/r10k-1.3.2/lib/r10k/task_runner.rb:21:in `run'
/usr/lib64/ruby/gems/2.0.0/gems/r10k-1.3.2/lib/r10k/cli/deploy.rb:67:in `block (2 levels) in command'
/usr/lib64/ruby/gems/2.0.0/gems/cri-2.5.0/lib/cri/command.rb:298:in `call'
/usr/lib64/ruby/gems/2.0.0/gems/cri-2.5.0/lib/cri/command.rb:298:in `run_this'
/usr/lib64/ruby/gems/2.0.0/gems/cri-2.5.0/lib/cri/command.rb:251:in `run'
/usr/lib64/ruby/gems/2.0.0/gems/cri-2.5.0/lib/cri/command.rb:264:in `run'
/usr/lib64/ruby/gems/2.0.0/gems/cri-2.5.0/lib/cri/command.rb:264:in `run'
/usr/lib64/ruby/gems/2.0.0/gems/r10k-1.3.2/bin/r10k:7:in `<top (required)>'
/usr/bin/r10k:23:in `load'
/usr/bin/r10k:23:in `<main>'
[R10K::Task::Deployment::PurgeEnvironments - INFO] Purging stale environments from /etc/puppet/environments
```

**Comment by adrienthebo** [comment]
This should be resolved in f8f8a8d1, thanks for reporting this!
